### PR TITLE
Introduce `enable-debug-assert` to enable/disable debug asserts at runtime

### LINKED ISSUE #569: [NEW] Activate `debugServerAssertWithInfo` using a server config
> "assert" that only fires during tests would totally work for me. Do you have an issue on it already?

This already exists today, it's `debugServerAssertWithInfo`. By default `debugServerAssertWithInfo` is compiled out, but it's compiled in for some tests. See `-DDEBUG_ASSERTIONS` like https://github.com/valkey-io/valkey/blob/fd58b73f0ae895bf9de3810d799da20bb75a2b4f/.github/workflows/daily.yml#L601C102-L601C120.

_Originally posted by @madolson in https://github.com/valkey-io/valkey/issues/526#issuecomment-2135847088_

### PR DESCRIPTION
**Pull request by PingXie**
Introduce a new server configuration, `enable-debug-assert`, which allows selectively enabling or disabling, at runtime, expensive or risky assertions used primarily for debugging and testing.

Fix #569

### CODE REVIEW

File: src/config.c [hunk 1]
```diff
@@ -3191,7 +3191,9 @@ standardConfig static_configs[] = {
     createTimeTConfig("repl-backlog-ttl", NULL, MODIFIABLE_CONFIG, 0, LONG_MAX, server.repl_backlog_time_limit, 60 * 60, INTEGER_CONFIG, NULL, NULL), /* Default: 1 hour */
     createOffTConfig("auto-aof-rewrite-min-size", NULL, MODIFIABLE_CONFIG, 0, LLONG_MAX, server.aof_rewrite_min_size, 64 * 1024 * 1024, MEMORY_CONFIG, NULL, NULL),
     createOffTConfig("loading-process-events-interval-bytes", NULL, MODIFIABLE_CONFIG | HIDDEN_CONFIG, 1024, INT_MAX, server.loading_process_events_interval_bytes, 1024 * 1024 * 2, INTEGER_CONFIG, NULL, NULL),
+    createBoolConfig("enable-debug-assert", NULL, MODIFIABLE_CONFIG, server.enable_debug_assert, 0, NULL, NULL),
```
Comment (enjoy-binbin) [inline src/config.c]: let's add a HIDDEN_CONFIG? I feel that this configuration item should not be exposed to users, that is, it should not be put into valkey.conf (or it should be a debug subcommand)

Comment (PingXie) [reply src/config.c]: make sense. will remove it from valkey.conf. 

> it should be a debug subcommand

I think a real (hidden) server config would give us coverage on the server initialization path (right after configs parsing) as well.

**Review by enjoy-binbin — APPROVED** [review]
let's also remove DEBUG_ASSERTIONS flag in ours ci yml (and daily.yml)

File: src/config.c [hunk 2]
```diff
@@ -3191,7 +3191,9 @@ standardConfig static_configs[] = {
     createTimeTConfig("repl-backlog-ttl", NULL, MODIFIABLE_CONFIG, 0, LONG_MAX, server.repl_backlog_time_limit, 60 * 60, INTEGER_CONFIG, NULL, NULL), /* Default: 1 hour */
     createOffTConfig("auto-aof-rewrite-min-size", NULL, MODIFIABLE_CONFIG, 0, LLONG_MAX, server.aof_rewrite_min_size, 64 * 1024 * 1024, MEMORY_CONFIG, NULL, NULL),
     createOffTConfig("loading-process-events-interval-bytes", NULL, MODIFIABLE_CONFIG | HIDDEN_CONFIG, 1024, INT_MAX, server.loading_process_events_interval_bytes, 1024 * 1024 * 2, INTEGER_CONFIG, NULL, NULL),
+    createBoolConfig("enable-debug-assert", NULL, MODIFIABLE_CONFIG | HIDDEN_CONFIG, server.enable_debug_assert, 0, NULL, NULL),
```
Comment (madolson) [inline src/config.c:3194]: Why is this here instead of with the other bool configs?

File: tests/assets/default.conf [hunk 3]
```diff
@@ -35,3 +35,5 @@ propagation-error-behavior panic
 
 # Make sure shutdown doesn't fail if there's an initial AOFRW
 shutdown-on-sigterm force
+
+enable-debug-assert yes
```
Comment (madolson) [inline tests/assets/default.conf:39]: This isn't sufficient, this is only used for normal tests but it won't cover the cluster tests. Maybe also update `init-tests.tcl`.

File: src/config.c (hunk 1 above)
Comment (madolson) [reply src/config.c]: Historically we've marked all configs as major decisions since it is a change to the user facing API. Users can toggle this config at runtime. By default, we recommend users should disable the `DEBUG` command. So, I think this should be marked as a static config so that nobody has the ability to enable this.

**Comment by zuiderkwast** [comment]
In #599 we rename this config `enable-debug-assert` → `debug-assert-enabled` to match other bool config names. It's OK to do this because this config hasn't been released yet.

**Comment by madolson** [comment]
> In https://github.com/valkey-io/valkey/pull/599 we rename this config enable-debug-assert → debug-assert-enabled to match other bool config names. It's OK to do this because this config hasn't been released yet.

Yeah, that should be OK.
